server/bundler/plugin.js

```diff
@@ -8,15 +8,20 @@ ChunkFileNames.prototype.apply = function( compiler ) {
 			return this.asString( [
 				"// \"0\" is the signal for \"already loaded\"",
 				"if ( installedChunks[ chunkId ] === 0 ) {",
-				this.indent("return callback.call( null, " + this.requireFn + " );"),
+				this.indent("return Promise.resolve();"),
 				"}",
-				"// an array means \"currently loading\".",
-				"if ( installedChunks[ chunkId ] !== undefined ) {",
-				this.indent( "installedChunks[ chunkId ].push( callback );" ),
+				"// a Promise means \"currently loading\".",
+				"if ( installedChunks[ chunkId ] ) {",
+				this.indent( "return installedChunks[ chunkId ][2];" ),
 				" } else { ",
 				this.indent( [
 					"// start chunk loading",
-					"installedChunks[ chunkId ] = [ callback ];",
+					"var promise = new Promise( function( resolve, reject ) {",
```

Comment:
ccing some people that were involved with the initial introduction in case they have input: @rralian @nb